test/task.cc

```diff
@@ -41,13 +52,149 @@ TEST(TaskTest, Succeed) {
 
   EXPECT_EQ(42, *e3());
 
-  auto e4 = [&]() {
-    return e3()
-        | Then([](int i) {
-             return i + 1;
-           })
-        | e3();
+  auto e4 = [&]() -> Task<int> {
+    return Task<int>([&]() {
+      return e3()
+          | Then([](int i) {
+               return i + 1;
+             })
+          | e3();
+    });
   };
 
   EXPECT_EQ(42, *e4());
 }
+
+TEST(Task, Void) {
+  int x = 0;
+  auto e1 = [&x]() -> Task<void> {
+    return Task<void>([&x]() {
+      return Then([&x]() {
+        x = 100;
+      });
+    });
+  };
+
+  *e1();
+
+  EXPECT_EQ(100, x);
+}
+
+TEST(Task, FailOnCallback) {
+  struct Functions {
+    MockFunction<void()> fail, stop, start;
+  };
+
+  Functions functions;
+
+  EXPECT_CALL(functions.fail, Call())
+      .Times(1);
+
+  EXPECT_CALL(functions.stop, Call())
+      .Times(0);
+
+  EXPECT_CALL(functions.start, Call())
+      .Times(0);
+
+  auto e = [&]() -> Task<int> {
+    return Task<int>([&]() {
+      return Eventual<int>()
+                 .context("error")
```

Comment:
Seems like you don't actually need the `.context("error")` here, can youjust do `k.Fail("error");` below please?